crs_motion_planning/src/freespace_planning_server.cpp

```diff
@@ -0,0 +1,192 @@
+#include <rclcpp/rclcpp.hpp>
+
+#include <rclcpp_action/rclcpp_action.hpp>
+#include <std_srvs/srv/trigger.hpp>
+#include <trajectory_msgs/msg/joint_trajectory.hpp>
+
+#include <crs_msgs/srv/call_freespace_motion.hpp>
+
+#include <tf2/transform_storage.h>
+#include <tf2/transform_datatypes.h>
+#include <tf2_eigen/tf2_eigen.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#include <tf2_ros/transform_listener.h>
+#include <tf2_ros/buffer.h>
+
+#include <tesseract_motion_planners/trajopt/trajopt_motion_planner.h>
+#include <tesseract_motion_planners/trajopt/config/trajopt_planner_default_config.h>
+#include <tesseract_motion_planners/trajopt/config/trajopt_planner_freespace_config.h>
+#include <tesseract_environment/core/environment.h>
+#include <tesseract_environment/core/utils.h>
+
+#include <trajopt/problem_description.hpp>
+
+#include <tesseract_rosutils/utils.h>
+
+#include <ament_index_cpp/get_package_share_directory.hpp>
+
+void tesseractRosutilsToMsg(trajectory_msgs::msg::JointTrajectory& traj_msg,
+                              const std::vector<std::string>& joint_names,
+                              const Eigen::Ref<const tesseract_common::TrajArray>& traj)
+{
+  assert(joint_names.size() == static_cast<unsigned>(traj.cols()));
+
+  // Initialze the whole traject with the current state.
+  std::map<std::string, int> jn_to_index;
+  traj_msg.joint_names.resize(joint_names.size());
+  traj_msg.points.resize(static_cast<size_t>(traj.rows()));
+
+  for (int i = 0; i < traj.rows(); ++i)
+  {
+    trajectory_msgs::msg::JointTrajectoryPoint jtp;
+    jtp.positions.resize(static_cast<size_t>(traj.cols()));
+
+    for (int j = 0; j < traj.cols(); ++j)
+    {
+      if (i == 0)
+        traj_msg.joint_names[static_cast<size_t>(j)] = joint_names[static_cast<size_t>(j)];
+
+      jtp.positions[static_cast<size_t>(j)] = traj(i, j);
+    }
+
+    jtp.time_from_start = rclcpp::Duration(i, 0);  // TODO: re-enable with rclcpp::Duration
+    traj_msg.points[static_cast<size_t>(i)] = jtp;
+  }
+}
+
+class PlanningServer: public rclcpp::Node
+{
+public:
+  PlanningServer()
+    : Node("planning_server_node")
+  {
+    plan_service_ = this->create_service<crs_msgs::srv::CallFreespaceMotion>("test_plan", std::bind(&PlanningServer::planService, this, std::placeholders::_1, std::placeholders::_2));
+    traj_publisher_ = this->create_publisher<trajectory_msgs::msg::JointTrajectory>("set_trajectory_test",10);
+    joint_state_listener_ = this->create_subscription<sensor_msgs::msg::JointState>("/joint_states", 2, std::bind(&PlanningServer::jointCallback, this, std::placeholders::_1));
+
+    joint_name_list_ = {"shoulder_pan_joint", "shoulder_lift_joint", "elbow_joint", "wrist_1_joint", "wrist_2_joint", "wrist_3_joint"};
```

Comment:
It'd be nice to get this list from a parameter or the RobotModel class if there's one in ROS2

Reply:
I now do a lookup through tesseract